plat/x86: Add `write-combined` page entry attribute

jschlumpp opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

This attribute is useful for memory-mapped devices and can reduce the amount of transactions necessary when communicating with the device.
<!--
Please provide a detailed description of the changes made in this new PR.
-->

jschlumpp committed:
plat/x86: Add `write-combined` page entry attribute

This attribute is useful for memory-mapped devices and can reduce the
amount of transactions necessary when communicating with the device.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

nderjung commented on plat/common/include/x86/paging.h:
```diff
@@ -266,6 +268,15 @@ pgarch_init(void)
+		return -ENOTSUP;
+	}
+	/* Write default value to MSR, in case BIOS modified them */
+	wrmsrl(0x277, 0x0007040600070406);
```
How/where is the default value provided to MSR determined, is there a reference you can cite? (as an "outsider", this value seems opaque to me)

jschlumpp replied:
Good point, I will add a reference to the comment.

marcrittinghaus commented on arch/x86/x86_64/include/uk/asm/paging.h:
```diff
@@ -193,6 +193,7 @@ static inline int ukarch_vaddr_range_isvalid(__vaddr_t start, __vaddr_t end)
 #define PAGE_ATTR_PROT_READ		0x01 /* Page is readable */
 #define PAGE_ATTR_PROT_WRITE		0x02 /* Page is writeable */
 #define PAGE_ATTR_PROT_EXEC		0x04 /* Page is executable */
+#define PAGE_ATTR_WRITECOMBINE		0x08 /* Allow write-combining on page*/
```
Shorten the comment so it fits better into the line.

```C
#define PAGE_ATTR_WRITECOMBINE		0x08 /* Page allows write-combining */
```

marcrittinghaus committed:
include/arch/x86_64: Add PAT bits

This commit adds definitions for configuration of the page attribute
table on x86.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
plat/common/x86: Add PAT MSR

This commit adds the definition of the page attribute table (PAT)
configuration MSR.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus commented on plat/common/include/x86/paging.h:
```diff
@@ -266,6 +269,18 @@ pgarch_init(void)
+		uk_pr_crit("Page table attributes are not supported.\n");
+		return -ENOTSUP;
+	}
+	/* Write default value described in Intel's Software Developer
+	 * Manual to the MSR, in case BIOS modified them.
+	 * (See Vol 3, 11.12.4 Programming the PAT)
+	 */
+	wrmsrl(0x277, 0x0007040600070406);
```
I agree with Alex here that this is still a bit cumbersome to read. Please cherry-pick the commits d7333a30816d7e419c6af13e716706e784c9f606 and a0695d6e00490876baef88a97ba6e2d2ab4b6d23 from this branch https://github.com/marcrittinghaus/unikraft/tree/pr_788_writecombine.

Then you can write:
```C
/* Reset PAT to default value */
wrmsrl(X86_MSR_PAT, X86_PAT_DEFAULT);
```

marcrittinghaus commented on plat/common/include/x86/paging.h:
```diff
@@ -98,13 +99,15 @@ pgarch_pte_create(__paddr_t paddr, unsigned long attr, unsigned int level,
+
 	/* Take all other bits from template */
 	pte |= template & (X86_PTE_US |
-			   X86_PTE_PWT |
-			   X86_PTE_PCD |
```
We should also remove these from the `pt_pte_create` variant then. Make sure to add a remark in the comment similar to the one for the protections.

marcrittinghaus reviewed (changes requested):
Hey @mschlumpp,

thanks for this! I have added commits for some definitions. Please include them in your PR.

Also make sure to adapt:
- `pgarch_pte_change_attr`
- `pgarch_attr_from_pte`

Otherwise, you loose the write combine attribute when splitting a large page.

marcrittinghaus reviewed (approved):
Works for me! All fine. Thanks @mschlumpp.

Reviewed-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

nderjung reviewed (approved):
Thanks!

Approved-by: Alexander Jung <alex@unikraft.io>
